File: Demos/11_AllPermutations_demo.py
def AllPermutations(nums):
    prevPermutations = [[]]
    allPermutations = []

    for num in nums:
        allPermutations = []
        for permutation in prevPermutations:
            for i in range(len(permutation) + 1):
                newPerm = permutation.copy()
                newPerm.insert(i, num)
                allPermutations.append(newPerm)
        
        prevPermutations = allPermutations

    allPermutations.sort()
    return allPermutations

if __name__ == "__main__":
    nums = [1,3,5]
    expectedOutput = [[1,3,5], [1,5,3], [3,1,5], [3,5,1], [5,1,3], [5,3,1]]
    output = AllPermutations(nums)
    print(output)
    print(expectedOutput)
    print(output == expectedOutput)
    print()

File: Demos/C++/11_AllPermutations_demo.cpp
#include<iostream>
#include<vector>

std::vector<std::vector<int>> AllPermutations(std::vector<int> &nums)
{
    std::vector<std::vector<int>> prevPermutations = {{}};
    std::vector<std::vector<int>> allPermutations = {};

    for (int num : nums)
    {
        allPermutations = {};
        for (std::vector<int> permutation : prevPermutations)
        {
            for (int i = 0; i < permutation.size() + 1; i ++)
            {
                std::vector<int> newPerm = permutation;
                newPerm.insert(newPerm.begin() + i, num);
                allPermutations.push_back(newPerm);
            }
        }
        prevPermutations = allPermutations;
    }

    return allPermutations;
}

void PrintVector(std::vector<int> nums)
{
    std::cout << "{ ";
    for (int num : nums)
    {
        std::cout << num << " ";
    }
    std::cout << "} ";
}

void PrintVectorOfVectors(std::vector<std::vector<int>> vectorOfVectors)
{
    for (std::vector<int> innerVector : vectorOfVectors)
    {
        PrintVector(innerVector);
    }
    std::cout << std::endl;
}

int main()
{
    std::vector<int> nums = {1, 2, 3};
    std::vector<std::vector<int>> output = AllPermutations(nums);

    std::cout << "Sample:" << std::endl;
    PrintVector(nums);

    std::cout << "\n\nAll permutations:" << std::endl;
    PrintVectorOfVectors(output);

    return 0;
}